tests/snapshot/snapshot_restore_k8s_upgrade_test.go

```diff
@@ -0,0 +1,154 @@
+package snapshot
+
+import (
+	"testing"
+
+	"github.com/gruntwork-io/terratest/modules/terraform"
+	"github.com/rancher/shepherd/clients/rancher"
+	management "github.com/rancher/shepherd/clients/rancher/generated/management/v3"
+	"github.com/rancher/shepherd/extensions/users"
+	password "github.com/rancher/shepherd/extensions/users/passwordgenerator"
+	ranchFrame "github.com/rancher/shepherd/pkg/config"
+	namegen "github.com/rancher/shepherd/pkg/namegenerator"
+	"github.com/rancher/shepherd/pkg/session"
+	"github.com/rancher/tfp-automation/config"
+	"github.com/rancher/tfp-automation/framework"
+	cleanup "github.com/rancher/tfp-automation/framework/cleanup"
+	"github.com/rancher/tfp-automation/tests/extensions/provisioning"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+)
+
+type SnapshotRestoreK8sUpgradeTestSuite struct {
+	suite.Suite
+	client             *rancher.Client
+	standardUserClient *rancher.Client
+	session            *session.Session
+	terraformConfig    *config.TerraformConfig
+	clusterConfig      *config.TerratestConfig
+	terraformOptions   *terraform.Options
+}
+
+func (s *SnapshotRestoreK8sUpgradeTestSuite) SetupSuite() {
+	testSession := session.NewSession()
+	s.session = testSession
+
+	client, err := rancher.NewClient("", testSession)
+	require.NoError(s.T(), err)
+
+	s.client = client
+
+	enabled := true
+	var testuser = namegen.AppendRandomString("testuser-")
+	var testpassword = password.GenerateUserPassword("testpass-")
+	user := &management.User{
+		Username: testuser,
+		Password: testpassword,
+		Name:     testuser,
+		Enabled:  &enabled,
+	}
+
+	newUser, err := users.CreateUserWithRole(client, user, "user")
+	require.NoError(s.T(), err)
+
+	newUser.Password = user.Password
+
+	standardUserClient, err := client.AsUser(newUser)
+	require.NoError(s.T(), err)
+
+	s.standardUserClient = standardUserClient
+
+	terraformConfig := new(config.TerraformConfig)
+	ranchFrame.LoadConfig(config.TerraformConfigurationFileKey, terraformConfig)
+
+	s.terraformConfig = terraformConfig
+
+	clusterConfig := new(config.TerratestConfig)
+	ranchFrame.LoadConfig(config.TerratestConfigurationFileKey, clusterConfig)
+
+	s.clusterConfig = clusterConfig
+
+	terraformOptions := framework.Setup(s.T())
+	s.terraformOptions = terraformOptions
+}
+
+func (s *SnapshotRestoreK8sUpgradeTestSuite) TestSnapshotRestoreK8sUpgrade() {
+	nodeRolesAll := []config.Nodepool{config.AllRolesNodePool}
+	nodeRolesShared := []config.Nodepool{config.EtcdControlPlaneNodePool, config.WorkerNodePool}
+	nodeRolesDedicated := []config.Nodepool{config.EtcdNodePool, config.ControlPlaneNodePool, config.WorkerNodePool}
+
+	snapshotRestoreK8sVersion := config.TerratestConfig{
+		SnapshotInput: config.Snapshots{
+			UpgradeKubernetesVersion: s.clusterConfig.UpgradedKubernetesVersion,
+			SnapshotRestore:          "kubernetesVersion",
+			RecurringRestores:        1,
+		},
+	}
+
+	snapshotRestoreAll := config.TerratestConfig{
```

Comment:
nit; is it necessary to define a new config.TerratestConfig object here, or could you just define the config.Snapshots instead?

Reply:
Yeah, it has to be here. I tried doing the latter first, but from how I refactored the config, you have to define the `config.TerratestConfig` first.